Mesh repairs in the mesh generation phase

GiudGiud opened the pull request:
~~Will get one or two mesh repair operations before merging~~

Refs #25107

Commit:
Add a repair mesh generator
- remove/merge overlapping nodes

Commit:
Work on tests for repairing meshes

Commit:
Mesh repair generator:
- add option to split using uniform refinements all elements above a certain size
refs #25107

GiudGiud committed:
Add option to flip negative volume elements,
in mesh repair generator
refs #25107

GiudGiud committed:
Add custom tolerance for merging overlapping nodes

GiudGiud committed:
Perform multiple rounds of element refinement for breaking large elements
Move big-elem refinement to the refineBlockGenerator

GiudGiud committed:
Add a utility to rename a subdomain in the mesh

GiudGiud committed:
Add the ability to split subdomains by element types
Re-organize tests
refs #25107

moosebuild commented:
Job [Documentation](https://civet.inl.gov/job/1680394/) on d4a0600 wanted to post the following:

View the site [here](https://mooseframework.inl.gov/docs/PRs/25088/site)

This comment will be updated on new commits.

moosebuild commented:
Job [Coverage](https://civet.inl.gov/job/1680412/) on d4a0600 wanted to post the following:


## Framework coverage

<table>
  <thead>
    <tr>
      <th></th>
      <th>db3b49</th>
      <th colspan=4>#25088 d4a060</th>
    </tr>
    <tr>
      <th></th>
      <th>Total</th>
      <th>Total</th>
      <th>+/-</th>
      <th>New</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td><b>Rate</b></td>
      <td>85.61%</td>
      <td>85.56%</td>
      <td>-0.05%</td>
      <td>28.57%</td>
    </tr>
    <tr>
      <td><b>Hits</b></td>
      <td>89696</td>
      <td>89716</td>
      <td>+20</td>
      <td>26</td>
    </tr>
    <tr>
      <td><b>Misses</b></td>
      <td>15074</td>
      <td>15143</td>
      <td>+69</td>
      <td>65</td>
    </tr>
  </tbody>
</table>

[Diff coverage report](https://mooseframework.inl.gov/docs/PRs/25088/coverage_diff/framework)

[Full coverage report](https://mooseframework.inl.gov/docs/PRs/25088/coverage/framework)


## Modules coverage

Coverage did not change

## Full coverage reports

<details>
<summary>Reports</summary>

- [`framework`](https://mooseframework.inl.gov/docs/PRs/25088/coverage/framework)
- [`chemical_reactions`](https://mooseframework.inl.gov/docs/PRs/25088/coverage/chemical_reactions)
- [`combined`](https://mooseframework.inl.gov/docs/PRs/25088/coverage/combined)
- [`contact`](https://mooseframework.inl.gov/docs/PRs/25088/coverage/contact)
- [`electromagnetics`](https://mooseframework.inl.gov/docs/PRs/25088/coverage/electromagnetics)
- [`external_petsc_solver`](https://mooseframework.inl.gov/docs/PRs/25088/coverage/external_petsc_solver)
- [`fluid_properties`](https://mooseframework.inl.gov/docs/PRs/25088/coverage/fluid_properties)
- [`fsi`](https://mooseframework.inl.gov/docs/PRs/25088/coverage/fsi)
- [`functional_expansion_tools`](https://mooseframework.inl.gov/docs/PRs/25088/coverage/functional_expansion_tools)
- [`geochemistry`](https://mooseframework.inl.gov/docs/PRs/25088/coverage/geochemistry)
- [`heat_conduction`](https://mooseframework.inl.gov/docs/PRs/25088/coverage/heat_conduction)
- [`level_set`](https://mooseframework.inl.gov/docs/PRs/25088/coverage/level_set)
- [`misc`](https://mooseframework.inl.gov/docs/PRs/25088/coverage/misc)
- [`navier_stokes`](https://mooseframework.inl.gov/docs/PRs/25088/coverage/navier_stokes)
- [`optimization`](https://mooseframework.inl.gov/docs/PRs/25088/coverage/optimization)
- [`peridynamics`](https://mooseframework.inl.gov/docs/PRs/25088/coverage/peridynamics)
- [`phase_field`](https://mooseframework.inl.gov/docs/PRs/25088/coverage/phase_field)
- [`porous_flow`](https://mooseframework.inl.gov/docs/PRs/25088/coverage/porous_flow)
- [`ray_tracing`](https://mooseframework.inl.gov/docs/PRs/25088/coverage/ray_tracing)
- [`rdg`](https://mooseframework.inl.gov/docs/PRs/25088/coverage/rdg)
- [`reactor`](https://mooseframework.inl.gov/docs/PRs/25088/coverage/reactor)
- [`richards`](https://mooseframework.inl.gov/docs/PRs/25088/coverage/richards)
- [`scalar_transport`](https://mooseframework.inl.gov/docs/PRs/25088/coverage/scalar_transport)
- [`solid_properties`](https://mooseframework.inl.gov/docs/PRs/25088/coverage/solid_properties)
- [`stochastic_tools`](https://mooseframework.inl.gov/docs/PRs/25088/coverage/stochastic_tools)
- [`tensor_mechanics`](https://mooseframework.inl.gov/docs/PRs/25088/coverage/tensor_mechanics)
- [`thermal_hydraulics`](https://mooseframework.inl.gov/docs/PRs/25088/coverage/thermal_hydraulics)
- [`xfem`](https://mooseframework.inl.gov/docs/PRs/25088/coverage/xfem)

</details>

## Warnings

- `framework` new line coverage rate 28.57% is less than the suggested 90.0%



This comment will be updated on new commits.

GiudGiud committed:
Prettier output of element types

GiudGiud committed:
Avoid considering nodes twice when removing overlapping nodes
refs #25107

GiudGiud committed:
Add tests for:
- overlapping nodes merging
- negative volume element flip
- mixed element type subdomain split
refs #25107

GiudGiud committed:
Fix test for single-region refinement
Add test for element-size driven block refinement
refs #22117

GiudGiud committed:
Add documentation for the mesh repair generator

GiudGiud committed:
Efficiency improvements for refine block
- avoid setting not prepared if did not refine
- avoid trying to refine with size if no too-small elements

GiudGiud committed:
No support for distributed meshes and repair operations

GiudGiud committed:
Apply size-based refinement on all processes together, or none

lindsayad commented on framework/src/utils/Conversion.C:
```diff
@@ -517,6 +517,92 @@ stringify(ElementType t)
+    case INFQUAD4:
+      return "INFQUAD4";
+    case INFQUAD6:
+      return "INFQUAD6";
+    case INFHEX8:
+      return "INFHEX8";
+    case INFHEX16:
+      return "INFHEX16";
+    case INFHEX18:
+      return "INFHEX18";
+    case INFPRISM6:
+      return "INFPRISM6";
+    case INFPRISM12:
+      return "INFPRISM12";
+    case NODEELEM:
+      return "NODEELEM";
+    case REMOTEELEM:
+      return "REMOTEELEM";
+    case TRI3SUBDIVISION:
+      return "TRI3SUBDIVISION";
+    case TRISHELL3:
+      return "TRISHELL3";
+    case QUADSHELL4:
+      return "QUADSHELL4";
+    case QUADSHELL8:
+      return "QUADSHELL8";
+    case TRI7:
+      return "TRI7";
+    case TET14:
+      return "TET14";
+    case PRISM20:
+      return "PRISM20";
+    case PRISM21:
+      return "PRISM21";
+    case PYRAMID18:
+      return "PYRAMID18";
+    default:
+      mooseError("unrecognized type");
+  }
+}
```
```suggestion
  return libMesh::Utility::enum_to_string(t);
}
```

GiudGiud replied:
haha

GiudGiud replied:
poor Daniel who I made to type this

lindsayad commented on framework/src/meshgenerators/TiledMeshGenerator.C:
```diff
@@ -67,7 +67,7 @@ TiledMeshGenerator::generate()
   std::unique_ptr<MeshBase> initial_mesh = std::move(_input);
   if (!initial_mesh->is_replicated())
-    mooseError("SmoothMeshGenerator is not implemented for distributed meshes");
+    mooseError("TiledMeshGenerator is not implemented for distributed meshes");
```
We could just omit the name since it comes out automatically in `MooseObject::mooseError`

GiudGiud replied:
```suggestion
    mooseError("This MeshGenerator is not implemented for distributed meshes");
```

lindsayad commented on framework/src/meshgenerators/MeshRepairGenerator.C:
```diff
@@ -0,0 +1,162 @@
+                                             "Name of the mesh generator providing the mesh");
+
+  params.addParam<bool>("fix_node_overlap", false, "Whether to merge overlapping nodes");
+  params.addParam<Real>("node_overlap_tol", 1e-8, "Tolerance for merging overlapping nodes");
```
Should state whether this is an absolute or relative type tolerance

GiudGiud replied:
```suggestion
  params.addParam<Real>("node_overlap_tol", 1e-8, "Absolute tolerance for merging overlapping nodes");
```

lindsayad commented on framework/src/meshgenerators/MeshRepairGenerator.C:
```diff
@@ -0,0 +1,162 @@
+  if (_fix_overlapping_nodes)
+  {
+    auto pl = mesh->sub_point_locator();
+    std::set<dof_id_type> nodes_removed;
```
```suggestion
    std::unordered_set<dof_id_type> nodes_removed;
```

lindsayad replied:
It seems like we should actually remove the nodes we no longer use

GiudGiud replied:
added a remove_oprhaned_nodes at the end

lindsayad commented on framework/src/meshgenerators/MeshRepairGenerator.C:
```diff
@@ -0,0 +1,162 @@
+        {
+          auto new_id = next_block_id + i++;
+          // Create blocks when a block has multiple element types
+          mesh->subdomain_name(new_id) = mesh->subdomain_name(id) + "_" + Moose::stringify(type_it);
```
```suggestion
          mesh->subdomain_name(new_id) = mesh->subdomain_name(id) + "_" + Moose::stringify(type);
```

lindsayad commented on framework/src/meshgenerators/MeshRepairGenerator.C:
```diff
@@ -0,0 +1,162 @@
+
+          // Re-assign elements to the new blocks
+          for (auto elem : mesh->active_subdomain_elements_ptr_range(id))
+            if (elem->type() == type_it)
```
```suggestion
            if (elem->type() == type)
```

lindsayad commented on framework/src/meshgenerators/RefineBlockGenerator.C:
```diff
@@ -73,7 +76,47 @@ RefineBlockGenerator::generate()
+      if (!_enable_neighbor_refinement)
+        refinedmesh.face_level_mismatch_limit() = 0;
+      refinedmesh.refine_elements();
+    }
+    i++;
```
```suggestion
      refined_on_size = true;
    }
```

lindsayad commented on framework/include/meshgenerators/RefineBlockGenerator.h:
```diff
@@ -37,6 +37,9 @@ class RefineBlockGenerator : public MeshGenerator
   /// Toggles whether neighboring level one elements should be refined or not. Defaults to true.
   const bool _enable_neighbor_refinement;
 
+  /// maximum element size over which elements are refined
+  const Real _max_element_size;
```
I would rename `size` to `volume`. `size` could be related to `hmin()` or `hmax()` for example, so there's some ambiguity

GiudGiud replied:
will do

GiudGiud commented on framework/doc/content/source/meshgenerators/MeshRepairGenerator.md:
```diff
@@ -0,0 +1,23 @@
+- flipping the orientation of negative volume elements. Negative volume elements can arise from a wrong orientation.
+
+- separating elements in subdomains into several subdomains depending on their element types. Subdomains with a mix of
+  element types are not supported by [Exodus.md] output. The new split subdomains have the type of the element appended.
```
```suggestion
  element types are not supported by [Exodus.md] output. The new split subdomains' names have the type of the element appended to their respective names.
```

GiudGiud commented on framework/doc/content/source/meshgenerators/MeshRepairGenerator.md:
```diff
@@ -0,0 +1,23 @@
+- flipping the orientation of negative volume elements. Negative volume elements can arise from a wrong orientation.
+
+- separating elements in subdomains into several subdomains depending on their element types. Subdomains with a mix of
+  element types are not supported by [Exodus.md] output. The new split subdomains have the type of the element appended.
```
better now?

lindsayad replied:
yep!

GiudGiud commented on framework/include/meshgenerators/RefineBlockGenerator.h:
```diff
@@ -37,6 +37,9 @@ class RefineBlockGenerator : public MeshGenerator
   const bool _enable_neighbor_refinement;
 
+  /// maximum element size over which elements are refined
+  const Real _max_element_size;
```
```suggestion
  const Real _max_element_volume;
```

GiudGiud commented on framework/src/meshgenerators/MeshRepairGenerator.C:
```diff
@@ -0,0 +1,162 @@
+  mesh->prepare_for_use();
+
+  // Blanket ban on distributed. This can be relaxed for some operations if needed
+  if (!mesh->is_replicated())
+    mooseError("MeshRepairGenerator is not implemented for distributed meshes");
```
is_serial feels replicated?
```
true if all elements and nodes of the mesh exist on the current processor, false otherwise
```

GiudGiud replied:
we do replicated elsewhere

roystgnr replied:
The docs try to explain the intent of each method: "if all elements and nodes of the mesh exist on the current processor" for `is_serial()`, "if new elements and nodes can and should be created in synchronization on all processors" for `is_replicated()`.

If you're trying to figure out whether you can read any existing element on any processor, `is_serial()` is the right thing to use.  If you're trying to figure out whether you can create local elements independently from processor to processor, `!is_replicated()` is.  But IIRC the latter is an older API so I'm sure (even in libMesh) it's still being used in a ton of places where `is_serial()` would have been more appropriate.

roystgnr replied:
For mesh generators, I'm currently a fan of `MeshSerializer`, to replace the "are you serial?" question with "YOU ARE SERIAL", but allowing future steps in mesh generation to re-parallelize.

This still isn't quite right, though.  `MeshSerializer` both serializes and (in its destructor) unserializes.  Wasteful if the mesh is soon going to need to be serialized again ... but the alternative of serializing and leaving it serialized to the end is wasteful if the mesh is going to go through processes (many generators) that are more efficient on a distributed mesh.  I'm not sure what the efficient way to handle this is.  Have mesh generators register whether they need serialized inputs or not, then distribute an output mesh as soon as it's no longer going to be an input to a subsequent generator?

lindsayad replied:
It seems like a follow-on mesh generator that knows it's very heavy and can work with a distributed mesh, could check `!is_replicated() && is_serial()` and act accordingly. It's perhaps best for the current mesh generator to not guess, do less operations, register its state (accessible through APIs like `is_serial()` and `is_prepared()`) and assume follow-on generators will do what's best for themselves.

GiudGiud replied:
Since this serial/replicated distinction paradigm has not made it to MOOSE, I can with either there but is_replicated seems like the more consistent treatment in MOOSE

GiudGiud replied:
ok found it in a few places

GiudGiud replied:
changed.

GiudGiud commented on framework/src/meshgenerators/MeshRepairGenerator.C:
```diff
@@ -0,0 +1,162 @@
+        {
+          for (auto & elem_node : elem->node_ref_range())
+          {
+            const Real tol = _node_overlap_tol;
```
it s to shorten the name

lindsayad replied:
Do your fingies get tired?

GiudGiud commented on test/tests/meshgenerators/refine_block_generator/test_single.i:
```diff
@@ -19,7 +19,7 @@
     type = RefineBlockGenerator
     input = eg
-    block = '0'
+    block = '3'
```
was probably playing around the input

GiudGiud commented on test/tests/meshgenerators/refine_block_generator/gold/test_single_in.e:
let s see if I can remove that

GiudGiud commented on test/tests/meshgenerators/refine_block_generator/test_single.i:
```diff
@@ -19,7 +19,7 @@
     type = RefineBlockGenerator
     input = eg
-    block = '0'
+    block = '3'
```
```suggestion
    block = '0'
```

GiudGiud committed:
Apply suggestions from code review

- use libmesh enum conversion
- prettier code, better loops
- better param description
- use is_serial to prepare for a future where that status might change dynamically
- use volume instead of size for element areas/volumes

Co-authored-by: Alex Lindsay <alexander.lindsay@inl.gov>

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/1699934/) on 5900658 wanted to post the following:

Your code requires style changes.  

A patch was auto generated and copied [here](https://mooseframework.inl.gov/docs/PRs/25088/clang_format/style.patch)  
You can directly apply the patch by running, in the top level of your repository:

`curl -s https://mooseframework.inl.gov/docs/PRs/25088/clang_format/style.patch | git apply -v`  

Alternatively, with your repository up to date and in the top level of your repository:

`git clang-format 3c86f3d37ab48d0f73eaeb6f3d5e025a00541269`

GiudGiud commented on test/tests/meshgenerators/refine_block_generator/gold/test_single_in.e:
oh no I cant remove that.
The initial test was doing nothing! refinement of 0

GiudGiud commented:
ugh mis-force pushed. This is what i get for naming my PRs close names

GiudGiud commented:
Distributed sweeps wasnt necessary btw, mesh repairs are only run on serial and tests are limited to replicated

GiudGiud commented:
Thanks for the review!

If there is interest, I m happy to make that generator work for distributed meshes. I think we need 1 level of ghosting that s it, for some of them, and a global name communication for another repair (the elem types in a block)

GiudGiud commented:
But I ll need some real funding. I did this on flex time :S

loganharbour commented:
Broke next https://civet.inl.gov/job/1721618/

lindsayad commented:
@GiudGiud if you want any help debugging the failed assertion, let me know

GiudGiud commented:
yeah no idea. How do I get _nodes to be refreshed and not have a nullptr in it?

GiudGiud commented:
now hitting this:
```
meshgenerators/mesh_repair_generator.node_overlap: *** ERROR ***
meshgenerators/mesh_repair_generator.node_overlap: 
meshgenerators/mesh_repair_generator.node_overlap: [1] ../src/mesh/mesh_tools.C, line 232, compiled Aug 16 2023 at 09:56:22
```
which has to do with parallel id counts
https://mooseframework.inl.gov/docs/doxygen/libmesh/mesh__tools_8C_source.html

I have this in the code and top branch is being run
```
    if (mesh->allow_renumbering())
    {
      // mesh->remove_orphaned_nodes();
      mesh->renumber_nodes_and_elements();
      // mesh->update_parallel_id_counts();
    }
    else
    {
      mesh->remove_orphaned_nodes();
      mesh->update_parallel_id_counts();
    }
```

roystgnr commented:
> which has to do with parallel id counts

In `assert_dofobj_unique_id`?  That's not *directly* a discrepancy in parallel id counts - it's screaming that two different processors see the same type of DofObject (either Elem or Node) with the same id() but differing unique_id() values.  That could be caused by `_next_unique_id` or `_next_unpartitioned_unique_id` being out of sync, but there are lots of other ways it can happen.

GiudGiud commented:
how do I get unique ids in sync? It seems update_parallel_id_counts() get _next_unique_id in sync.
How about _next_unpartitioned_unique_id ?

roystgnr commented:
`update_parallel_id_counts()` gets both, when both exist - `_next_unpartitioned_unique_id` is only for the `DistributedMesh` case, where processors might be doing any combination of adding new unpartitioned DofObjects in tandem with each other and adding new local DofObjects independently of each other.

roystgnr commented:
> resulting in a disagreement about what node should be removed

That would be a good example of the "lots of other ways it could happen".  Even if the nodes were in the exact same place and we gave them the exact same connectivity, if we tried to treat them as the same node but they had different unique ids then we'd scream about that during a `dbg` mode check.

GiudGiud commented:
ah yes this is it.
I meant for all these loops to NOT be local. And there is still one, this one, on local_node_ptr_range
